### PR TITLE
dx: add typescript-eslint

### LINKED ISSUE #154: [DX] add typescript lint and remove unnecessary file pattern in eslintignore/prettierignore
### Description

# defaang - configuration improvement plan

I'll list out what I want to do in this issue. If you have any suggestions, please let me know.

## tsconfig.json

Add 3 additional type checking.

### include

As mentioned in [tsconfig#include]:

> If a glob pattern doesn’t include a file extension, then only files with supported extensions are included (e.g. .ts, .tsx, and .d.ts by default, with .js and .jsx if allowJs is set to true).

We could remove file extension in option `include`. And add `"*"` to include root path config files.

### noImplicitReturns

Enable error reporting for codepaths that do not explicitly return in a function.

See more: [tsconfig#noImplicitReturns]

### noUnusedLocals

Enable error reporting when a local variables aren't read.

See more: [tsconfig#noUnusedLocals]

### noUnusedParameters

Raise an error when a function parameter isn't read

See more: [tsconfig#noUnusedParameters]

~~### noImplicitThis~~ (This is include in strict options)

~~Enable error reporting when this is given the type any.~~

~~See more: [tsconfig#noImplicitThis]~~

~~### noImplicitAny~~ (This is include in strict options)

~~Enable error reporting for expressions and declarations with an implied any type.~~

~~See more: [tsconfig#noImplicitAny]~~

[tsconfig#include]: https://www.typescriptlang.org/tsconfig#include
[tsconfig#noimplicitreturns]: https://www.typescriptlang.org/tsconfig#noImplicitReturns
[tsconfig#noimplicitthis]: https://www.typescriptlang.org/tsconfig#noImplicitThis
[tsconfig#noimplicitany]: https://www.typescriptlang.org/tsconfig#noImplicitAny
[tsconfig#nounusedlocals]: https://www.typescriptlang.org/tsconfig#noUnusedLocals
[tsconfig#nounusedparameters]: https://www.typescriptlang.org/tsconfig#noUnusedParameters

## package.json

Add `type-check` script to check typescript errors using `tsc --noEmit`

## .eslintrc.json

Add plugin `@typescript-eslint/eslint-plugin` and its parser `@typescript-eslint/parser` for typescript lint.

### @typescript-eslint/eslint-plugin

An ESLint plugin which provides lint rules for TypeScript codebases.

npm module: [@typescript-eslint/eslint-plugin](https://www.npmjs.com/package/@typescript-eslint/eslint-plugin)

### @typescript-eslint/parser

An ESLint parser which leverages TypeScript ESTree to allow for ESLint to lint TypeScript source code.

Set `parserOptions.project` to `./tsconfig.json`. Provide `tsconfig`'s rules for the parser.

npm module: [@typescript-eslint/parser](https://www.npmjs.com/package/@typescript-eslint/parser)

## .eslintignore

According to [eslint ignoring code]:

> In addition to any patterns in the .eslintignore file, ESLint always follows a couple of implicit ignore rules even if the --no-ignore flag is passed. The implicit rules are as follows:
>
> 1. node_modules/ is ignored.
> 2. dot-files (except for .eslintrc.\*), as well as dot-folders and their contents, are ignored.

We could remove both `node_modules` and `dot-files` in the `.eslintignore`. And also, remove unnecessary files like: `/.pnp` and the debug files to keep it clean.

These are what I think should be removed:

- node_modules
- /pnp
- .pnp.js
- /build
- .DS_Store
- \*.pem
- npm-debug.log\*
- yarn-debug.log\*
- yarn-error.log\*
- .pnpm-debug.log\*
- .env\*.local
- .vercel
- \*.tsbuildinfo

[eslint ignoring code]: https://eslint.org/docs/latest/user-guide/configuring/ignoring-code

## .prettierignore

Same as ESlint, by default, Prettier ignores node_modules (if [--with-node-modules CLI option][1] not specified).  
Prettier only [works for specific file extensions][2]. Therefore, we could remove some files in `.prettierignore`.

These are what I think should be removed:

- node_modules
- /pnp
- .pnp.js
- /build
- .DS_Store
- \*.pem
- npm-debug.log\*
- yarn-debug.log\*
- yarn-error.log\*
- .pnpm-debug.log\*
- .env\*.local
- .vercel
- \*.tsbuildinfo

[1]: https://prettier.io/docs/en/cli.html#--with-node-modules
[2]: https://prettier.io/docs/en/index.html

### PR DESCRIPTION
**Pull request by benny123tw**
### 🛠️ Fixes Issue

Closes #154  

### 👨‍💻 Changes proposed

* add `@typescript-eslint/eslint-plugin` and `@typescript-eslint/parser` to ESLint
* add 5 type check to `tsconfig`
* add type check script to `package.json` to check Typescript errors
* remove unnecessary pattern in `prettierignore` and `eslintignore`

### 📄 Note to reviewers

none

### 📷 Screenshots

none

### CODE REVIEW

**Comment by iShibi** [comment]
Thank you for your contribution. Unfortunately, we can't merge these changes because of the following reasons:

- The linter in Next.js already supports TypeScript out of the box.
- The `strict` rule in `tsconfig` turns on all the `strict mode family options`, no need to turn them on individually again.
- The contents of the `.ignore` files are not doing any harm, so let's not change something that's not broken.

**Comment by benny123tw** [comment]
Hi @iShibi, thanks for the comment.

1. I don't think [eslint-plugin-next](https://github.com/vercel/next.js/blob/canary/packages/eslint-plugin-next/lib/index.js) has Typescript lint.
2. didn't notice strict mode is on, haha, but strict mode only enables all **strict type-checking options**, which do not include **Additional Checks** like `noUnusedLocals`, `noUnusedParameters`, or `noImplicitReturns`.
3. removing unnecessary patterns in the `.ignore` file can keep things clean and not break any things, IMO.

Just a suggestion. It's all up to you.

File: package.json [hunk 1]
```diff
@@ -4,6 +4,7 @@
 		"dev": "cross-env NODE_OPTIONS='--inspect' next dev",
 		"build": "next build",
 		"start": "next start",
+		"type-check": "tsc --noEmit",
```
Comment (iShibi) [inline package.json:7]: Just to keep it consistent with other npm scripts name
```suggestion
		"type:check": "tsc --noEmit",
```

File: tsconfig.json [hunk 2]
```diff
@@ -1,20 +1,26 @@
 {
 	"compilerOptions": {
-		"target": "es5",
+		"target": "esnext",
```
Comment (iShibi) [inline tsconfig.json:3]: Since this file was generated by Next.js itself, we should keep the target to `es5` for now, just to be safe.
```suggestion
		"target": "es5",
```

File: tsconfig.json [hunk 3]
```diff
@@ -1,20 +1,26 @@
...
 		"esModuleInterop": true,
-		"module": "esnext",
 		"moduleResolution": "node",
 		"resolveJsonModule": true,
 		"isolatedModules": true,
 		"jsx": "preserve",
-		"incremental": true
+		"incremental": true,
+		"noImplicitReturns": true,
+		"noImplicitThis": true,
```
Comment (iShibi) [inline tsconfig.json:18]: This is `true` if strict is set. You can remove this one.
```suggestion
```

File: tsconfig.json [hunk 4]
```diff
@@ -1,20 +1,26 @@
...
+		"noImplicitAny": true,
```
Comment (iShibi) [inline tsconfig.json:19]: Same
```suggestion
```

File: tsconfig.json [hunk 5]
```diff
@@ -1,20 +1,26 @@
...
+		"strictNullChecks": true,
```
Comment (iShibi) [inline tsconfig.json:20]: Same
```suggestion
```

File: tsconfig.json [hunk 6]
```diff
@@ -1,20 +1,26 @@
...
+		"noUnusedLocals": true,
+		"noUnusedParameters": true
 	},
-	"include": ["next-env.d.ts", "**/*.ts", "**/*.tsx"],
+	"include": ["next-env.d.ts", "src/**/*", "src/**/*", "*"],
```
Comment (iShibi) [inline tsconfig.json:24]: Can you revert this change? This was also generated by Next.js itself and is fine as it is.

**Review by iShibi — CHANGES_REQUESTED** [review]
Hey @benny123tw. After going through the changes again, I think we can use some of them. I have added comments to changes which I think are not required. For the `.ignore` files I still believe we should keep them as it is.

File: tsconfig.json (hunk 6 above)
Comment (benny123tw) [reply tsconfig.json]: Hi @iShibi, thanks for taking the time to review this. I reverted this change but kept`"*"` to include config files in the root path, or the parser will raise an error. (or put them in ignore files)

**Comment by iShibi** [comment]
Looks good. Now to pass the vercel build, run `npm run type:check` and fix the errors that show up. Most of them are unused variables that you can just remove. In `hello.ts` replace `req` with an `_`

**Comment by benny123tw** [comment]
> Looks good. Now to pass the vercel build, run `npm run type:check` and fix the errors that show up. Most of them are unused variables that you can just remove. In `hello.ts`  replace `req` with an `_`

Done. Should we put `npm run type:check` in the commit hook? Because I found that it is hard to debug on Vercel deploy (to someone that is no permission).

**Comment by iShibi** [comment]
I'll add that in a PR I was thinking about creating to add a type check job in the test workflow.

**Review by iShibi — APPROVED** [review]
